File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c is ')' or c is '}' or c is ']':
                if not stack:
                    return False
                if c is ')':
                    if stack[-1] is '(':
                        stack.pop()
                    else:
                        return False
                if c is '}':
                    if stack[-1] is '{':
                        stack.pop()
                    else:
                        return False
                if c is ']':
                    if stack[-1] is '[':
                        stack.pop()
                    else:
                        return False
            else:
                stack.append(c)
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        
        for(int i = 0; i< s.length(); i++){
            if((s.charAt(i) == '(' ) || (s.charAt(i) == '{') || (s.charAt(i) == '[')){
                stack.push(s.charAt(i));
            }
            else if (!stack.isEmpty()){ 
                if((s.charAt(i) == ')' && stack.peek() == '(') 
                ||  (s.charAt(i) == '}' && stack.peek() == '{')
                ||  (s.charAt(i) == ']' && stack.peek() == '[')){
                    stack.pop();
                    continue;
                }
            }
            if((s.charAt(i) == ')' ) || (s.charAt(i) == '}') || (s.charAt(i) == ']')){
                return false;
            }
        }
        
        return stack.isEmpty();
    }
}
